feat: add receipts

wilsonianb opened the pull request:
https://github.com/interledger/rfcs/issues/568

Depends on https://github.com/interledger/rfcs/pull/570

Implementation details include:
- the server encodes the receipt nonce and secret (from SPSP queries) and appends them together as an additional segment after the token/connection tag in the destination address
- the latest receipt is exposed to the sender on the stream object allowing them to get it with every `outgoing_money` event and/or after calling `sendTotal`

wilsonianb commented on src/connection.ts:
```diff
@@ -1115,6 +1136,15 @@ export class Connection extends EventEmitter {
+          if (frame.type === FrameType.StreamReceipt) {
+            const stream = this.streams.get(frame.streamId.toNumber())
+            if (stream) {
+              stream.emit('receipt', frame.receipt)
```
Should receipts instead be emitted in the `outgoing_money` event so that sender's can more easily verify receipt amounts?

kincaidoneil commented on src/server.ts:
```diff
@@ -121,11 +129,30 @@ export class Server extends EventEmitter {
+   * @param receiptSecret Optional secret to use for signing STREAM receipts
    */
-  generateAddressAndSecret (connectionTag?: string): { destinationAccount: string, sharedSecret: Buffer } {
+  generateAddressAndSecret (opts?: string | GenerateAddressSecretOpts): { destinationAccount: string, sharedSecret: Buffer } {
```
If the receipt nonce and secret are appended as a separate segment, it'd be easier for intermediaries to identify, remove or alter it (is that a concern?). Also, any logic elsewhere that only uses the last segment of the ILP address to identify a connection may work incorrectly after this change

What if the connection ID, tag, receipt nonce and secret were all encoded into a single encrypted blob/token so they can't be separately tracked or modified?

wilsonianb replied:
I actually had the nonce and secret encoded with the tag initially (https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/82981060a513138d7cc2d4c76fc714f24d480a4f) but changed it (https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/757ccff5ce39858abbc0541092caba6fdb18bbad).

I think I do like the idea of making them part of the connection token (like the connection tag is currently) in order to make tampering detectable.

kincaidoneil replied:
The connection tag and receipt nonce are still opaque/can be tampered with or removed. Instead of only encrypting the receipt secret using the server secret -- what if the whole last segment of the address (id, tag, receipt nonce, receipt secret) was encrypted using the server secret?

wilsonianb replied:
I think they're tamper-proof since they're part of the id/token used to generate the shared secret and verified at the start of the connection: 
https://github.com/interledgerjs/ilp-protocol-stream/blob/d2c6b14fd2ff7266b37adbc46c45834f2d55de6f/src/pool.ts#L93-L109
If that's correct, then I was thinking still just encrypt the receipt secret to keep the destination address from being even loooonger.

kincaidoneil replied:
Ah, gotcha -- yeah, that should be tamper proof.

> If that's correct, then I was thinking still just encrypt the receipt secret to keep the destination address from being even loooonger.

How much longer would it be if it was encrypted?

One idea to shorten it: could we use the receipt nonce as the ID for the connection instead of randomly generating another token?

wilsonianb replied:
Actually it looks like the encryption used always adds an extra 28 bytes. (I was previously scared off by the fact that the receipt secret size went from 32 bytes to 60 bytes encrypted.)
So I can encode and encrypt the token, connection tag, receipt nonce, and receipt secret all together.

I think the STREAM server should still generate its own ID instead of relying on unique receipt nonces in the (proxied) SPSP query.

wilsonianb commented on src/connection.ts:
```diff
@@ -1111,6 +1132,17 @@ export class Connection extends EventEmitter {
 
       if (responsePacket.ilpPacketType === IlpPacketType.Fulfill) {
+        for (let frame of responsePacket.frames) {
+          if (frame.type === FrameType.StreamReceipt) {
```
:thinking: I'm not sure
On the plus side, detecting an invalid receipt (by checking for consistent nonce and accurate id, amount & stream start time) could allow the sender to immediately stop sending money.
OTOH:
- the sender can't verify the hmac
- this STREAM implementation isn't currently tracking the total amount received by the counterparty for a stream (we're just taking their word for it:
https://github.com/interledgerjs/ilp-protocol-stream/blob/476c12daf215fbb7573174de1351f4cc1d78a13d/src/connection.ts#L733-L739

kincaidoneil replied:
On the connection level, if the packet is fulfilled, the sender can & should assume that at least the minimum destination amount from the Prepare got delivered to the recipient ([this line](https://github.com/interledgerjs/ilp-protocol-stream/blob/681cd740f5f619994c01f8922f734eedaba50b65/src/connection.ts#L1151) should probably be changed to `addTotalDelivered(max(requestPacket.prepareAmount, responsePacket.prepareAmount))`). Likewise, the amount of the receipt could be validated against the cumulative amount we know got delivered to the recipient.

_But_, because receipt accounting is on a per-stream basis rather than per-connection, it might be complicated to compute the amount delivered on each stream

kincaidoneil replied:
Is it a concern if the receiver underreports the amount they receive to the verifier? Or is it assumed that the verifier is trusting the receiver to honestly report the amount they receive?

wilsonianb replied:
I think the sender is the only who stands to lose anything from an underreported receipt amount.
Unless there was some scenario where the verifier was confirming that a certain amount *hasn't* been paid, but that doesn't seem doable with depending on the sender to get receipts to the verifier.

wilsonianb commented on package.json:
```diff
@@ -43,6 +43,7 @@
   },
   "dependencies": {
     "@types/node": "^10.14.22",
+    "base64url": "^3.0.1",
```
I opted for the package since I also needed to decode base64url (https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/757ccff5ce39858abbc0541092caba6fdb18bbad#diff-916156c481141bfcd08c6ae40ce10e00R219) which isn't as concise.
Assuming I'll still need to decode for how we end encoding the receipt secret and key in the destination address (https://github.com/interledgerjs/ilp-protocol-stream/pull/133#discussion_r395326204), I can plan on adding base64url decode functionality instead of using the package.

sentientwaffle replied:
Isn't it just:

```
function decodeBase64 (string) {
  return Buffer.from(string
      .replace(/-/g, '+')
      .replace(/_/g, '/'),
    'base64')
}
```

wilsonianb replied:
ah, you're right. I was looking at the library's implementation without thinking.
All the more reason not to use it.

sentientwaffle commented on test/connection.test.ts:
```diff
@@ -520,6 +527,30 @@ describe('Connection', function () {
+      }
+
+      assert.calledTwice(spy)
+      assert.calledWith(spy.firstCall, await createReceipt(this.receiptNonce, this.receiptSecret, '500'))
```
Maybe add an assertion that compares the receipt to a fixed (hard-coded) value.

wilsonianb replied:
updated it to use the fixture: https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/96d786dd5b62ab81ea3280876d4a6bcfdb4bfc9c

wilsonianb commented:
squashed (full commit history available [here](https://github.com/wilsonianb/ilp-protocol-stream/commits/og-receipts)) and rebased
Added https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/95e3882b880a4e0a9eae0b46705e6283c89edadb and https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/17fc6015558301b4437eae5e8bc345334bb435c4

wilsonianb commented:
The spec is now merged merged (https://github.com/interledger/rfcs/pull/570). This is now as ready for review as ever.

sharafian commented on src/pool.ts:
```diff
@@ -109,3 +140,10 @@ export class ServerConnectionPool {
 }
+
+function decodeBase64url (str: string) {
+  return Buffer.from(str
```
node's Buffer.from accepts base64url as valid base64

wilsonianb replied:
story checks out
https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings
updated
https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/59ba7c63f7ecff28e22c14d460deb1ae1f9d5795

sharafian commented on src/util/receipt.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+}
+
+export function verifyReceipt (receipt: Buffer, secret: Buffer): Boolean {
```
we might want something like the JWT library's decode, which validates and also returns the deserialized fields

wilsonianb replied:
Updated `verifyReceipt` to return decoded fields like ([JWT's verify](https://github.com/auth0/node-jsonwebtoken#jwtverifytoken-secretorpublickey-options-callback))
https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/7468b13d2b4d4e4044004c727509d1131e5d70d1

sharafian commented on src/util/crypto-node.ts:
```diff
@@ -58,3 +66,15 @@ export function generateSharedSecretFromToken (seed: Buffer, token: Buffer): Buf
   return sharedSecret
 }
+
+export function generateReceiptHMAC (secret: Buffer, receipt: Buffer): Buffer {
```
calling this buffer a receipt is a little confusing because a full receipt contains the hmac

wilsonianb replied:
renamed to `message`
https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/7856838aa567b26812c49e7c41484f4c092ae9bf

sharafian commented on test/connection.test.ts:
```diff
@@ -520,6 +528,23 @@ describe('Connection', function () {
+    it('should get a receipt for each fulfilled packet', async function () {
+      const clientStream = this.clientConn.createStream()
+      const spy = sinon.spy(clientStream, '_setReceipt')
+      await clientStream.sendTotal(1002)
```
should this also check that the receipt is exposed from the client stream?

wilsonianb replied:
added check
https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/66e008e12f951cc1a0ec2bd04f293f6a0b027a41
